### PR TITLE
Move image queue to use promises, move getArrayBuffer to use promises

### PR DESCRIPTION
**Pull request by HarelM**
The PR #3276 and its branch have too many changes and I can't seem to be able to fix the tests, I need to take slower steps still...
This is now ready for review

### CODE REVIEW

File: src/source/raster_dem_tile_source.ts [hunk 1]
```diff
@@ -58,16 +57,15 @@ export class RasterDEMTileSource extends RasterTileSource implements Source {
...
                 tile.state = 'unloaded';
                 callback(null);
-            } else if (err) {
-                tile.state = 'errored';
-                callback(err);
-            } else if (img) {
-                if (this.map._refreshExpiredTiles) tile.setExpiryData(expiry);
+            } else if (response && response.data) {
+                const img = response.data;
+                if (this.map._refreshExpiredTiles && response.cacheControl && response.expires) tile.setExpiryData({cacheControl: response.cacheControl, expires: response.expires});
```
Comment (neodescis) [inline src/source/raster_dem_tile_source.ts:68]: This got pretty long... maybe switch to `{ }` after the if?

File: src/source/raster_tile_source.ts [hunk 2]
```diff
@@ -158,20 +158,18 @@ export class RasterTileSource extends Evented implements Source {
...
             if (tile.aborted) {
                 tile.state = 'unloaded';
                 callback(null);
-            } else if (err) {
-                tile.state = 'errored';
-                callback(err);
-            } else if (img) {
-                if (this.map._refreshExpiredTiles && expiry) tile.setExpiryData(expiry);
+            } else if (response && response.data) {
+                if (this.map._refreshExpiredTiles && response.cacheControl && response.expires) tile.setExpiryData({cacheControl: response.cacheControl, expires: response.expires});
```
Comment (neodescis) [inline src/source/raster_tile_source.ts:168]: This one also got pretty long

File: src/source/raster_dem_tile_source.ts (hunk 1 above)
Comment (HarelM) [reply src/source/raster_dem_tile_source.ts:68]: Sounds right, thanks for the input!

File: src/source/raster_tile_source.ts (hunk 2 above)
Comment (HarelM) [reply src/source/raster_tile_source.ts:168]: Yes, it's the same thing... I'll fix both.

File: src/util/util.ts [hunk 3]
```diff
@@ -482,16 +454,16 @@ export function isImageBitmap(image: any): image is ImageBitmap {
  * ArrayBuffers.
  *
  * @param data - Data to convert
- * @param callback - A callback executed after the conversion is finished. Invoked with error (if any) as the first argument and resulting image bitmap (when no error) as the second
+ * @returns - A  promise executed after the conversion is finished. Invoked with error (if any) as the first argument and resulting image bitmap (when no error) as the second
```
Comment (neodescis) [inline src/util/util.ts:457]: Suggest changing "executed after" to "resolved when"

File: src/util/util.ts [hunk 4]
```diff
@@ -505,21 +477,23 @@ const transparentPngUrl = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAA
  * @param data - Data to convert
  * @param callback - A callback executed after the conversion is finished. Invoked with error (if any) as the first argument and resulting image element (when no error) as the second
```
Comment (neodescis) [inline src/util/util.ts:478]: Update
```suggestion
 * @param callback - A promise resolved when the conversion is finished. Invoked with error (if any) as the first argument and resulting image element (when no error) as the second
```

File: src/util/image_request.ts [hunk 5]
```diff
@@ -115,58 +108,55 @@ export namespace ImageRequest {
...
      */
-    export const getImage = (
-        requestParameters: RequestParameters,
-        callback: GetImageCallback,
-        supportImageRefresh: boolean = true
-    ): ImageRequestQueueItem => {
-        if (webpSupported.supported) {
-            if (!requestParameters.headers) {
-                requestParameters.headers = {};
+    export const getImage = (requestParameters: RequestParameters, abortController: AbortController, supportImageRefresh: boolean = true): Promise<GetResourceResponse<HTMLImageElement | ImageBitmap | null>> => {
```
Comment (neodescis) [inline src/util/image_request.ts:111]: Update doc comments?

**Comment by neodescis** [comment]
Looks good... just noticed some doc comments that should be updated

**Comment by HarelM** [comment]
Thanks! I'll look into those as well.

**Comment by HarelM** [comment]
Done, let me know if I can merge to the main branch.

File: src/util/image_request.ts [hunk 6]
```diff
@@ -95,78 +93,70 @@ export namespace ImageRequest {
...
     };
 
     /**
      * Request to load an image.
      * @param requestParameters - Request parameters.
-     * @param callback - Callback to issue when the request completes.
+     * @param abortController - allows to abort the request.
      * @param supportImageRefresh - `true`, if the image request need to support refresh based on cache headers.
-     * @returns Cancelable request.
+     * @returns - A promise resolved when the is returned.
```
Comment (neodescis) [inline src/util/image_request.ts:109]:
```suggestion
     * @returns - A promise resolved when the image is loaded.
```

File: src/util/util.ts [hunk 7]
```diff
@@ -482,16 +454,16 @@ export function isImageBitmap(image: any): image is ImageBitmap {
  * ArrayBuffers.
  *
  * @param data - Data to convert
- * @param callback - A callback executed after the conversion is finished. Invoked with error (if any) as the first argument and resulting image bitmap (when no error) as the second
+ * @returns - A  promise resolved when the conversion is finished. Invoked with error (if any) as the first argument and resulting image bitmap (when no error) as the second
```
Comment (neodescis) [inline src/util/util.ts:457]:
```suggestion
 * @returns - A  promise resolved when the conversion is finished
```

Sorry, I think I forgot to remove the now-irrelevant stuff in my suggestion

**Review by neodescis — APPROVED** [review]

**Comment by neodescis** [comment]
Go ahead and merge!

**Pull request merged**